src/NumericalAlgorithms/LinearSolver/LinearSolver.hpp

```diff
@@ -0,0 +1,242 @@
+  PreconditionedLinearSolver& operator=(PreconditionedLinearSolver&&) = default;
+
+  explicit PreconditionedLinearSolver(
+      std::optional<PreconditionerType> local_preconditioner) noexcept
+      : preconditioner_(std::move(local_preconditioner)) {}
+
+  // Override copy constructors so they can clone abstract preconditioners
+  PreconditionedLinearSolver(const PreconditionedLinearSolver& rhs) noexcept
+      : Base(rhs) {
+    if constexpr (not std::is_same_v<Preconditioner, NoPreconditioner>) {
+      preconditioner_ = rhs.clone_preconditioner();
+    }
+  }
+  PreconditionedLinearSolver& operator=(
+      const PreconditionedLinearSolver& rhs) noexcept {
+    Base::operator=(rhs);
+    if constexpr (not std::is_same_v<Preconditioner, NoPreconditioner>) {
+      preconditioner_ = rhs.clone_preconditioner();
+    }
+    return *this;
+  }
+
+ public:
+  ~PreconditionedLinearSolver() override = default;
+
+  /// \cond
+  explicit PreconditionedLinearSolver(CkMigrateMessage* m) noexcept : Base(m) {}
+  /// \endcond
+
+  void pup(PUP::er& p) noexcept override {  // NOLINT
+    PUP::able::pup(p);
+    if constexpr (not std::is_same_v<Preconditioner, NoPreconditioner>) {
+      p | preconditioner_;
+    }
+  }
+
+  /// Whether or not a preconditioner is set
+  bool has_preconditioner() const noexcept {
+    if constexpr (not std::is_same_v<Preconditioner, NoPreconditioner>) {
+      return preconditioner_.has_value();
+    } else {
+      return false;
+    }
+  }
+
+  // @{
+  /// Access to the preconditioner. Check `has_preconditioner()` before calling
+  /// this function. Calling this function when `has_preconditioner()` returns
+  /// `false` is an error.
+  template <bool Enabled = not std::is_same_v<Preconditioner, NoPreconditioner>,
+            Requires<Enabled> = nullptr>
+  const Preconditioner& preconditioner() const noexcept {
+    ASSERT(has_preconditioner(),
+           "No preconditioner is set. Please use `has_preconditioner()` to "
+           "check before trying to retrieve it.");
+    if constexpr (std::is_abstract_v<Preconditioner>) {
+      return **preconditioner_;
+    } else {
+      return *preconditioner_;
+    }
+  }
+
+  template <bool Enabled = not std::is_same_v<Preconditioner, NoPreconditioner>,
+            Requires<Enabled> = nullptr>
+  Preconditioner& preconditioner() noexcept {
+    ASSERT(has_preconditioner(),
+           "No preconditioner is set. Please use `has_preconditioner()` to "
+           "check before trying to retrieve it.");
+    if constexpr (std::is_abstract_v<Preconditioner>) {
+      return **preconditioner_;
+    } else {
+      return *preconditioner_;
+    }
+  }
+
+  /// Copy the preconditioner. Useful when it has an abstract type.
+  template <bool Enabled = not std::is_same_v<Preconditioner, NoPreconditioner>,
+            Requires<Enabled> = nullptr>
+  std::optional<PreconditionerType> clone_preconditioner() const noexcept {
```

Comment:
Why not name this `get_clone`? The reason should be documented since I'd like to avoid even more clone functions (we use `get_clone` because Charm defines a clone function that returns a raw (owning) pointer). Another comment: the Requires should also check that `not std::is_same_v<Preconditioner, NoPreconditioner>` because someone could define the template parameter explicitly and then chaos

Reply:
This is only a utility function to help implement `get_clone` in derived classes. Moved to `protected` and added details to the docs.

Reply:
Sorry, I don't understand what you mean with `derived classes implement get_clone`. Who and where will the definition be? I'm worried that this will be an ODR violation: you have multiple different definitions for the same function.